src/box/memtx_space.c

```diff
@@ -340,14 +341,42 @@ memtx_space_replace_tuple_common(struct space *space, struct txn_stmt *stmt,
 {
 	struct memtx_space *memtx_space = (struct memtx_space *)space;
 	struct tuple *result;
-	if (memtx_space->replace(space, old_tuple, new_tuple, mode,
-				 &result) != 0)
-		return -1;
+	struct tuple *orig_new_tuple = new_tuple;
+	bool was_referenced = false;
+	if (new_tuple != NULL && space->format->is_compressed) {
+		new_tuple = memtx_tuple_compress(new_tuple);
+		if (new_tuple == NULL)
+			return -1;
+		tuple_ref(new_tuple);
+		was_referenced = true;
+	}
+	int rc = memtx_space->replace(space, old_tuple, new_tuple,
+				      mode, &result);
+	if (rc != 0)
+		goto finish;
 	txn_stmt_prepare_rollback_info(stmt, result, new_tuple);
 	stmt->engine_savepoint = stmt;
-	stmt->new_tuple = new_tuple;
+	stmt->new_tuple = orig_new_tuple;
 	stmt->old_tuple = result;
-	return 0;
+	if (stmt->old_tuple != NULL) {
+		struct tuple *orig_old_tuple = stmt->old_tuple;
+		stmt->old_tuple = tuple_maybe_decompress(stmt->old_tuple);
+		if (stmt->old_tuple == NULL)
+			return -1;
```

Comment:
Shouldn't you clear `txn_stmt` (`new_tuple`, `engine_savepoint`) on failure? IMO better set them only on success.

Reply:
We should set `engine_savepoint` immidiatly after `replace` fucntion if we fail during allocation later we rollback this statement (see `memtx_engine_rollback_statement`).